src/lib/cloud/digitalocean/provisioner.ts

```diff
@@ -33,13 +31,89 @@ const REDIRECT_URIS: [string] = [
   //  'http://localhost:10101'
 ];
 
+const STORAGE_KEY_OAUTH = 'DigitalOcean-OAuth';
+
 interface KeyPair {
   private: string;
   public: string;
 }
 
 class Provisioner {
-  constructor(private dispatch_: Function, private state_: any = {}) {}
+  constructor(
+    private sshKeyPair_ :any = {},
+    private network_ :any = {},
+    private storage_ = freedom['core.storage'](),
+
+    // TODO: give this concept a better name, or use a different abstraction?
+    private cloud_ :any = null,
+
+    // doOAuth_() may be called multiple times concurrently before an OAuth
+    // flow has been completed. In order to ensure that each call doesn't
+    // trigger an additional OAuth flow in a new tab, the promise returned by
+    // the first call is cached in this field, and this cached promise is
+    // returned for subsequent calls.
+    private promiseOAuth_ :Promise<Object> = null
+  ) {}
+
+  /*
+   * Returns whether a DO OAuth token is saved in storage.
+   */
+  public hasOAuth = () : Promise<boolean> => {
+    return new Promise((F, R) => {
+      this.getOAuthFromStorage_().then(
+        (result) => { F(!!result); },
+        () => { F(false); }
+      );
+    });
+  }
+
+  /*
+   * TODO: Automatically exchange expired oauth tokens for fresh ones?
+   * Looks like this is supported:
+   * https://developers.digitalocean.com/documentation/oauth/#refresh-token-flow
+   * ref: https://github.com/uProxy/uproxy/issues/2565
+   */
+  private getOAuthFromStorage_ = () :Promise<any> => {
+    return new Promise((F, R) => {
+      let reject = (errcode :string, message :string, data? :any, exc? :any) => {
+        let err = {errcode: errcode, message: message, data: data, exc: exc};
+        console.debug(message, data);
```

Comment:
It turns out I wasn't seeing my log.debug() calls because I didn't realize there was a `consoleFilter` (i.e. log level) advanced setting I had to crank down first (see also #2599); had nothing to do with the (apparently spurious) freedom errors I was seeing about the logginglistener module. Reverted the s/log/console changes in 0f93bf9. Thanks @trevj!